repo/blob/throttling/throttling_storage.go

```
@@ -0,0 +1,97 @@
+	ReturnUnusedDownloadBytes(ctx context.Context, numBytes int64)
+}
+
+// throttlingStorage.
+type throttlingStorage struct {
+	blob.Storage
+	throttler Throttler
+}
+
+func (s *throttlingStorage) GetBlob(ctx context.Context, id blob.ID, offset, length int64, output blob.OutputBuffer) error {
+	acquired := length
+	if acquired < 0 {
+		acquired = unknownBlobAcquireLength
+	}
+
+	s.throttler.BeforeOperation(ctx, "GetBlob")
+	s.throttler.BeforeDownload(ctx, acquired)
+
+	output.Reset()
+
+	err := s.Storage.GetBlob(ctx, id, offset, length, output)
+	downloaded := int64(output.Length())
+
+	if acquired != downloaded {
+		if downloaded > acquired {
+			// we downloaded more than initially acquired, acquire more which may pause for a bit.
+			s.throttler.BeforeDownload(ctx, downloaded-acquired)
+		} else {
+			// we downloaded less than initially acquired, release extra
+			s.throttler.ReturnUnusedDownloadBytes(ctx, acquired-downloaded)
+		}
+	}
+
+	return err // nolint:wrapcheck
+}
+
+func (s *throttlingStorage) GetMetadata(ctx context.Context, id blob.ID) (blob.Metadata, error) {
+	s.throttler.BeforeOperation(ctx, "GetMetadata")
+
+	return s.Storage.GetMetadata(ctx, id) // nolint:wrapcheck
+}
+
+func (s *throttlingStorage) ListBlobs(ctx context.Context, blobIDPrefix blob.ID, cb func(bm blob.Metadata) error) error {
+	s.throttler.BeforeOperation(ctx, "ListBlobs")
+	return s.Storage.ListBlobs(ctx, blobIDPrefix, cb) // nolint:wrapcheck
+}
+
+func (s *throttlingStorage) SetTime(ctx context.Context, id blob.ID, t time.Time) error {
+	s.throttler.BeforeOperation(ctx, "SetTime")
+	return s.Storage.SetTime(ctx, id, t) // nolint:wrapcheck
+}
+
+func (s *throttlingStorage) PutBlob(ctx context.Context, id blob.ID, data blob.Bytes) error {
+	s.throttler.BeforeOperation(ctx, "PutBlob")
+	s.throttler.BeforeUpload(ctx, int64(data.Length()))
+
+	return s.Storage.PutBlob(ctx, id, data) // nolint:wrapcheck
```

nit; if PutBlob() fails early then we may sometimes not return the unused upload quota. It might help to have PutBlob() return the number of bytes written but I think this is okay being taxing on errors because that is not a happy-path. Plus, it may not be possible to get the uploaded bytes through every SDK.